Improved the results page for end of course for teachers.

henkka14 opened the pull request:
Currently the Results page for teachers is very minimal and quite frankly, useless. Tasked by Juho and Jaakko, I have been making some improvements to it. Added functionality to export the table data. Option to select from the exercise pool. Option to filter based on tags. Option to see different data indicators.

NOTE: Filtering relies on my other pull requests. I would also hope someone could test this with more student data. I have not found a way to test this reliably on large student result data.

Mankro commented on exercise/static/exercise/results_staff.js:
```diff
@@ -0,0 +1,840 @@
+let _maxPTrue = false;
+
+$("#total-subm-checkbox").change(function() {
+    _totalSubmTrue = this.checked ? true : false;
```
`_totalSubmTrue = this.checked ? true : false;`
Is this any different from simply assigning like this `_totalSubmTrue = this.checked;`

henkka14 replied:
It is same. Currently using `_totalSubmTrue = this.checked;`

raphendyr commented on exercise/templates/exercise/staff/results.html:
```diff
@@ -10,45 +11,185 @@
+<link rel="stylesheet" href="{% static 'exercise/css/results_staff.css' %}" />
+<script src="{% static 'exercise/results_staff.js' %}"
+        allExercisesUrl="/api/v2/exercises/"
+        exercisesUrl="/api/v2/courses/{{course.id}}/exercises"
```
These require url reverse resolving. No hardcoding urls in the files :)

henkka14 replied:
How would I find the url reversers for them all? I can only find url for course api

`courseUrl="{% url 'api:course-detail' version=2 course_id=course.id %}"`

And still would have to do `courseUrl + "students"`

Mankro replied:
I think you need to find the view class in the code and look for some URL key parameter that becomes the URL name for the URL resolver (maybe like this https://github.com/Aalto-LeTech/a-plus/blob/master/exercise/api/views.py#L66). The name is used in the static template tag.

In Python code, there is the function `api_reverse` that likewise uses the URL names to reverse them. 
https://github.com/Aalto-LeTech/a-plus/blob/master/aplus/api/__init__.py#L4

raphendyr replied:
Basenames are here: https://github.com/Aalto-LeTech/a-plus/blob/master/api/urls_v2.py . Url name is `api:basename-[list|detail]` depending if it's the list or detail version of the view handler.

```
exercisesUrl="{% url 'api:course-exercises-list' version=2 course_id=course.id %}"
studentsUrl="{% url 'api:course-students-list' version=2 course_id=course.id %}"
```

and so on

henkka14 replied:
Changed these.

raphendyr commented on locale/fi/LC_MESSAGES/django.po:
```diff
@@ -3485,3 +3485,74 @@ msgstr "Unohda automaattinen uudelleenohjaus."
 msgstr "Ei tietosuojailmoitusta. Ota yhteys ylläpitäjään!"
+
+
+#: results.html
```
Use `./manage.py makemessages -i 'venv/*'` to update this file.

henkka14 replied:
Done. However, there are some translations that are basically the same in `django.po` and `teacher.fi.json`. Would there be a clean way to use `_()` notation inside the html files?

raphendyr replied:
One day, teacher.fi.json will be generated from django.po or so. Untill then, we need to copy translations two two files.

raphendyr commented on exercise/templates/exercise/staff/results.html:
```diff
@@ -10,45 +11,185 @@
+            <!--Loading animation-->
+    <div class="row">
+        <div class="col-md-12">
+            <div id="loading-animation" class="loader">
```
The class is probably a bit generic (`loader`)? Also, can we use the same css for this and exercise loading? Should they have the same? Should we actually use this one for the exercises?

henkka14 replied:
changed the name to loading-animation. Did not do anything else yet. Not sure what would be good option.

raphendyr commented on exercise/staff_views.py:
```diff
@@ -179,7 +179,12 @@ class AllResultsView(CourseInstanceBaseView):
         super().get_common_objects()
         self.table = ResultTable(self.instance)
-        self.note("table")
+        self.tags = list(self.instance.usertags.all())
```
Follow #391 for how to do tags better now. Internal and external are handled via the same interface now.

henkka14 replied:
Changed the tags to include  USERTAG_EXTERNAL, USERTAG_INTERNAL.

raphendyr commented on exercise/static/exercise/results_staff.js:
```diff
@@ -0,0 +1,840 @@
+let currentScript = (document.currentScript ?
```
Missing the anonymous function scope wrapping!

henkka14 replied:
I assume I want to wrap the whole script?

henkka14 replied:
Actually I dont want to do that, but i am unsure how I would handle these variables I use in functions.

raphendyr replied:
Wrap the whole file and export functions that really need it. You can set variables to global scope with `window.variable = foo`. `window` is the global scope in browsers.

henkka14 replied:
I used this to do it. My question is how should I indent the file. Should I have one whole indent for the rest of the file just for the wrap function as I have in the current revision?

raphendyr replied:
It's good now

raphendyr commented on exercise/static/exercise/results_staff.js:
```diff
@@ -0,0 +1,840 @@
+    exerciseSelectionChange();
+});
+
+$("#avg-subm-checkbox").change(function() {
```
I would prefer mostly classes here. E.g. `$(this).find('input.results.avg-submissions').change(...`

henkka14 replied:
Currently using `$("input.avg-subm-checkbox").change(function() {`

raphendyr replied:
Works as long as there is only one result table (or table with those class names) on a single page. So, probably untill it's rewritten.

raphendyr commented on exercise/static/exercise/results_staff.js:
```diff
@@ -0,0 +1,840 @@
+        sumValue += value[0];
+    });
+
+// Empty data cells for total and tags in data indicators
```
Why these are not indented?

henkka14 replied:
Some style guide recommended putting all comments starting from left margin. Not sure what I should follow.

raphendyr replied:
sounds pretty interesting style guide. this seemed pretty nice https://github.com/airbnb/javascript  with exceptions on those features that are not supported by ie 11.

henkka14 replied:
https://www.crockford.com/code.html

This was the style guide I looked at for comments. I assume the rest are same as in airbnb guide aside from comments, which I changed.

raphendyr commented on exercise/static/exercise/results_staff.js:
```diff
@@ -0,0 +1,840 @@
+
+        let tag_ids = [];
+        _taggings.results.forEach(function(tagging) {
+        if (tagging.user.id === sid) {
```
indentation!

raphendyr commented on exercise/static/exercise/results_staff.js:
```diff
@@ -0,0 +1,840 @@
+        });
+
+        const tags = tag_ids.map(function (id) {
+        return id.toString();
```
there are apparenly multiple indentation problems here :(

raphendyr commented on exercise/templates/exercise/staff/results.html:
```diff
@@ -10,45 +11,185 @@
+        usersUrl="/api/v2/users/"
+        usertagsUrl="/api/v2/courses/{{course.id}}/usertags"
+        taggingsUrl="/api/v2/courses/{{course.id}}/taggings"
+        externalLabel="{{ external_user_label|safe }}"
```
Are these needed still?

Mankro replied:
If they are needed, they should be data attributes

henkka14 replied:
These were not needed

raphendyr commented on exercise/static/exercise/results_staff.js:
```diff
@@ -0,0 +1,840 @@
+    _students.results.forEach(function(student) {
+        let sid = student.id;
+        $.ajax({
+            url: usersUrl + sid
```
`url: student.url`, should work here, I presume.

raphendyr commented on exercise/static/exercise/results_staff.js:
```diff
@@ -0,0 +1,840 @@
+        module.exercises.forEach(function(exercise) {
+            let eid = exercise.id;
+            $.ajax({
+                url: allExercisesUrl + eid
```
`exercise.url` should have this value?

henkka14 commented:
Please review the current changes. Not making modifications currently unless until next comments.

raphendyr commented on exercise/static/exercise/results_staff.js:
```diff
@@ -0,0 +1,783 @@
+                let studentTags = student.tag_slugs;
+    
+                // Sort tags in order. Put internal and external tags in the beginning.
+                studentTags.sort(function(left, right){ 
```
This should be handled, if you keep the order from the API.

henkka14 replied:
Yep, seems to work without the sorting.

raphendyr commented on exercise/templates/exercise/staff/results.html:
```diff
@@ -10,45 +11,175 @@
+        data-usertags-url="{% url 'api:course-usertags-list' version=2 course_id=course.id %}"
+        data-taggings-url="{% url 'api:course-taggings-list' version=2 course_id=course.id %}"
+        data-points-url="{% url 'api:course-points-list' version=2 course_id=course.id %}"
+        data-external-label="{{ external_user_label|safe }}"
```
I don't think these are used anymore. I don't think they are provided by the view either anymore.

henkka14 replied:
Currently retrieving exercises, students, usertags and points URLs from API.

raphendyr commented on locale/fi/LC_MESSAGES/django.po:
```diff
@@ -3511,3 +3592,17 @@ msgstr "Unohda automaattinen uudelleenohjaus."
 #: userprofile/views.py:80
 msgid "No privacy notice. Please notify administration!"
 msgstr "Ei tietosuojailmoitusta. Ota yhteys ylläpitäjään!"
+
```
Add `--no-obsolete` to `makemessages`. Old messages are tracked by the git, so we don't need them in the file. They are used for translation programs, so they can find similar translations. Or if you remove and then add a string back.

henkka14 replied:
Done

raphendyr commented on locale/fi/LC_MESSAGES/django.po:
```diff
@@ -2623,23 +2624,97 @@ msgstr ""
+msgstr "Tehtävän maksimipisteet."
+
+#: exercise/templates/exercise/staff/results.html:117
+msgid "Max points"
```
Should this be `Maximum points` as it was before?

henkka14 replied:
Changed to the suggestions

raphendyr commented on exercise/templates/exercise/staff/results.html:
```diff
@@ -10,45 +11,175 @@
+
+    </div>
+
+        <!--Data table-->
```
Indentation starting from this line, doesn't seem to be consistent. Are there mix of tabs and spaces? I prefer tabs fro html files btw.

henkka14 replied:
Should be all tabs now the whole file, please note I use vs code for indentation that has its own customizations available. Feel free to check that all works in some other ide.

raphendyr commented on exercise/static/exercise/results_staff.js:
```diff
@@ -0,0 +1,783 @@
+        $(document.currentScript) :
+        $('script').last()); // Ugly solution for IE11
+
+    let exercisesUrl = currentScript.data("exercisesUrl");
```
These probably should be `const`.

henkka14 replied:
Changed some variables to const.

raphendyr commented on exercise/templates/exercise/staff/results.html:
```diff
@@ -10,45 +11,175 @@
+        data-points-url="{% url 'api:course-points-list' version=2 course_id=course.id %}"
+        data-external-label="{{ external_user_label|safe }}"
+        data-internal-label="{{ internal_user_label|safe }}"
+        defer>         
```
If I understand right, this means that the script is executed after html is parsed, but before document is ready (before `DOMContentLoaded`). All other scripts are hooked to jQuery document ready, which is run after `DOMContentLoaded` event (https://github.com/jquery/jquery/blob/cf84696fd1d7fe314a11492606529b5a658ee9e3/src/core/ready.js#L77). This includes translations. Thus, the script in question is always run before translation, which is kind of fine. That said, translations actually has a code that moves itself as a last document ready function, so that all other functions have time to register translations-ready listeners.

So, the defer vs. document ready here mostly is just different as anything else in A+. but technically it should be the same (at least when considering translations). In addition, as javascript runs all scripts in a single thread, there is no benefit from multi-threading.

So, I guess this is fine, just wrote the analysis down.

raphendyr commented:
Looks pretty good. Some small comments. In addittio, a lot of variables, which could be marked as `const`, are now `let`. It would be a good practice to always use `const` by default, and replace it with `let` if really needed. Typically, there is no need for `let` and most of algorithms using `let` can be written to be clearer with `const`s.

Mankro commented:
Has anyone tested whether this works correctly with unofficial or unconfirmed points? (unofficial = submissions after the deadline or max submissions limit if the category has the "allow unofficial" setting; unconfirmed = exercise points before the mandatory feedback exercise has been completed if the feedback exercise has the "confirm the level" setting)

Since the points are retrieved from the API, it might work correctly anyway, but it is worth checking.

raphendyr commented:
> Since the points are retrieved from the API, it might work correctly anyway, but it is worth checking.

Yeah it should work. API _should_ pull data from points cache, which should implement this logic for all different views.

henkka14 commented:
Also feel free to check if all looks good. Changed some things based on last comments. Also corrected a few minor issues, how `exercise` is found from `_points` and also fixed issue with modules that had no exercises, which gave NaN for their points.

raphendyr commented on locale/fi/LC_MESSAGES/django.po:
```diff
@@ -2108,6 +2113,18 @@ msgstr "Palautusta arvostellaan..."
+#, fuzzy
+#| msgid "Teacher of the course"
+msgid "Score on this course"
+msgstr "Kurssin opettaja"
```
I don't think  the automatic translation went so well :)

The `#, fuzzy` means that this was automatically translated from a string that was pretty close. Using a translation program helps to track which translations are final. Any case, check translations which include `fuzzy`, fix them and remove the two lines (fuzzy and source msgid).

raphendyr commented on locale/fi/LC_MESSAGES/django.po:
```diff
@@ -1131,48 +1132,52 @@ msgstr ""
-msgid "Course: {}, Instance: {}, not found."
-msgstr "Kurssi: {}, Instanssi: {}, ei löytynyt."
+#: edit_course/operations/configure.py:576
+msgid "Course URL '{}' doesn't match             service's hostname '{}'"
```
Dman. These are actually from previous changes, which did not update translations. I'll try to update them and then you can rebase again, so we get rid of them here.

raphendyr commented:
Master needs to update translations first, then we need to check this again. I think it's pretty much good. The unofficial points should be tested though.

On a side note, I don't personally like the method for building html, but I guess it's the simplest solution without better tools. So I guess we go with that.

raphendyr commented:
Master updated with translations for a missing set. Can you rebase and fix translations file.

henkka14 commented:
Rebased and fixed translations.

henkka14 commented:
Unofficial submissions are not accounted for, so if there are 7 submissions total where 5 are official and 2 are unofficial, the data indicators would only account for the 5 official submissions.

With the uncorfirmed points I am not sure, since I do not know how they are supposed to work. If you look at the picture below, the feedback form has the confirm_the_level option selected, yet it still gives points for the first exercise?

![confirm_the_level](https://user-images.githubusercontent.com/33518486/63044049-15d3ed80-bed6-11e9-8fea-b87c7ed555c7.jpg)

henkka14 commented:
Also note that assistants will be listed as students in the Results table if the assistant has enrolled to the course. I was under the assumption that this is not a problem, since assistants should not enroll to the course. However, I assume that the system allows assistants to enroll to the course if they want for whatever reason? Not sure exactly though

raphendyr commented on exercise/templates/exercise/staff/results.html:
```diff
@@ -10,45 +11,176 @@
+			</table>
+		</div>
+
+    </div>
```
Indent

raphendyr commented on exercise/templates/exercise/staff/results.html:
```diff
@@ -10,45 +11,176 @@
+						</select>
+					</td>
+				</tr>
+				</tbody>
```
Indent

raphendyr commented on exercise/templates/exercise/staff/results.html:
```diff
@@ -10,45 +11,176 @@
+<link rel="stylesheet" href="{% static 'exercise/css/loading_animation.css' %}" />
+<link rel="stylesheet" href="{% static 'exercise/css/results_staff.css' %}" />
+<script src="{% static 'exercise/results_staff.js' %}"
+		data-exercises-url="{% url 'api:course-exercises-list' version=2 course_id=course.id %}"
```
One tab would be sufficient

raphendyr commented:
> However, I assume that the system allows assistants to enroll to the course if they want for whatever reason? Not sure exactly though

The teacher can also enroll to the course :)

raphendyr commented:
> Thus, it also shows 200 points from 3.1.1. in the Results table.

How does it show module and A, B, C points? If those are correct, then I think it's good.

Mankro commented:
About the unconfirmed points (confirm the level): it is an awkward feature. It may require that the feedback questionnaire is in the same chapter as the exercises. Then it should require that the points of the chapter become active only after the feedback has been submitted and the "exercise results" page should be clear about it. The point badges are gray and the mouse hover text says something about mandatory feedback.

henkka14 commented:
> > Thus, it also shows 200 points from 3.1.1. in the Results table.
> 
> How does it show module and A, B, C points? If those are correct, then I think it's good. Of course, it would be good if the table would highlight that "these points are not actually counted, but as you reaquested to show exercise points, I did".

Uncorfirmed points do not work for any of the point categories, since the API always accepts the points in my testing, even if the feedback form is not submitted as in the picture above. Isn't it in the same chapter in the picture?

Anyway, fixed the indentation

Mankro commented:
I have a feeling that the A+ API does not know about unconfirmed points and it uses them immediately as normal points. If that is true, then only the user interface in the exercise results page really knows about the unconfirmed points feature. (It is only a hunch since I haven't tested it now.) Fixing/changing this is not part of this PR and I doubt it is urgent anyway since it has worked this way for a couple of years.

raphendyr commented:
I added `official` boolean to exercise entry in the api (in points). The boolean is true, if these exercise points are used for the grade (e.g. they are calculated to the module sum of points).

```json
   "modules": [
        {
            "id": 1,
            "max_points": 67,
            "points": 0,
            "points_by_difficulty": {
                "": 0
            },
            "exercises": [
                {
                    "id": 10,
                    "max_points": 2,
                    "points": 2,
                    "passed": true,
                    "official": false
```

Grade page could a) include button to select if unofficial points are shown and effecting summary calculations or b) calculate those as 0, but show the unofficial points in brackets (`0 (2)`).

For the teacher, it might be relevant to be able to see that some students are not getting as good grade as they have unconfirmed points (aka. unofficial due to not being confirmed).

Changes pushed to master.

raphendyr commented:
> raphendyr closed this now

miss click :(

Mankro commented:
The way you mixed the words "unofficial" and "unconfirmed" was confusing since they are separate features and I think the API supports unofficial points, but unconfirmed points maybe not.

raphendyr commented:
> The way you mixed the words "unofficial" and "unconfirmed"

They are used to refer to those specific features. Teacher would like to see unconfirmed points, but the API now exposes only combination of unofficial and unconfirmed. Though, only case when the `official = False` is true for a exe4rcise, which is confirmed, is when you submitted it after the deadline, but the points are confirmed.

Thus, we can show unofficial points, when teacher is looking for unconfirmed points as they should most of the time be the same thing. In addition, it's just good that the teacher will check the case when a student has submitted exercise after the DL.

And ofr the last part, the first submission to a exercise must be after the dl, otherwise the ready submission is used (submission is "official", when it's ready and confirmed).

raphendyr commented:
> I think the API supports unofficial points, but unconfirmed points maybe not.

API did expose ponts as long as they were not unofficial, but it didn't indicate in any way if they were unconfirmed. This was per exercise. For modules, API exposed the points values calculated internally, which did care if the points were unconfirmed or not.

Since this new page recalculates the summaries, it included also the unconfirmed points. Hence, I added the `official` flag to the API, so e.g. this page can filter per exercise points out, if they are not official points (included in the module points).

henkka14 commented:
> I added `official` boolean to exercise entry in the api (in points). The boolean is true, if these exercise points are used for the grade (e.g. they are calculated to the module sum of points).
> 
> ```json
>    "modules": [
>         {
>             "id": 1,
>             "max_points": 67,
>             "points": 0,
>             "points_by_difficulty": {
>                 "": 0
>             },
>             "exercises": [
>                 {
>                     "id": 10,
>                     "max_points": 2,
>                     "points": 2,
>                     "passed": true,
>                     "official": false
> ```
> 
> Grade page could a) include button to select if unofficial points are shown and effecting summary calculations or b) calculate those as 0, but show the unofficial points in brackets (`0 (2)`).
> 
> For the teacher, it might be relevant to be able to see that some students are not getting as good grade as they have unconfirmed points (aka. unofficial due to not being confirmed).
> 
> Changes pushed to master.

I used this logic to show them as (`0 (2)`).

henkka14 commented:
Rebased with master. I figured out my issue with confirm the level feedback form. Apparently it only works if the feedback gives points, so I override the feedback form to have 2 `max_points` in `conf.py`. It is a bit hacky for sure, but works I guess

I did use the `official` flag to see unconfirmed points inside parentheses. However, using the `official` flag did not let me get the unofficial points. For example, exercise with 5 submissions, if student first gets 5 submissions with 0 poins and with unofficial submission he scores 100 points, currently the option to show unofficial points would still show 0 points.

I could of course go through each exercise `submission` directly and check their `official` status and keep track of the highest point, but this would make the page much slower as the amount of ajax calls will be large.

Perhaps there could be `unofficial_points` in the `points` in API?

raphendyr commented:
> I used this logic to show them as `0 (2)`.

How that works with excel export? Can it drop the `(2)` part out from the excel, thus the cell would be plain number?

> Apparently it only works if the feedback gives points, so I override the feedback form to have 2 max_points in conf.py. It is a bit hacky for sure, but works I guess

Yep. This is how it works. Zero points can't effect anything as it means excercise is optional.

> If student first gets 5 submissions with 0 points and with unofficial submission he scores 100 points, currently the option to show unofficial points would still show 0 points.

> Perhaps there could be unofficial_points alongside the points in API?

I think current behavior is fine for now. Best probably would be to have best official and best unofficial submissions separately. We need to think that later if it seems to be something teachers would want. Though, I presume if the student has better unofficial score, they will contact staff via email and request them to be nice. On the other hand, there has been problems with unconfirmed points as people have missed the feedback often.

henkka14 commented:
> > I used this logic to show them as `0 (2)`.
> 
> How that works with excel export? Can it drop the `(2)` part out from the excel, thus the cell would be plain number?
> 
> > Apparently it only works if the feedback gives points, so I override the feedback form to have 2 max_points in conf.py. It is a bit hacky for sure, but works I guess
> 
> Yep. This is how it works. Zero points can't effect anything as it means excercise is optional.
> 
> > If student first gets 5 submissions with 0 points and with unofficial submission he scores 100 points, currently the option to show unofficial points would still show 0 points.
> 
> > Perhaps there could be unofficial_points alongside the points in API?
> 
> I think current behavior is fine for now. Best probably would be to have best official and best unofficial submissions separately. We need to think that later if it seems to be something teachers would want. Though, I presume if the student has better unofficial score, they will contact staff via email and request them to be nice. On the other hand, there has been problems with unconfirmed points as people have missed the feedback often.

It is very easy to ignore whole `<tr>` and `<td>` tags in the table. However, since the ` (2)` is a `<span>` tag inside `<td>` tag, it would be harder. Would mean that I would have to write own row for the exported table as hidden and one row for the display of the table, ie duplicate all rows. Seems a bit unnecessary at this point, since teacher can still just unclick the checkbox?

raphendyr commented on assets/js-translations/teacher.fi.json:
```diff
@@ -1,9 +1,104 @@
+
+  "Average submissions per student with submissions":
+  "Palautusten keskiarvo per opiskelija, jolla on vähintään yksi palautus",
+  
```
There is whitespace here (not visible in github :( )

raphendyr commented:
Can you run `sed -i 's/\s*$//' <file>` for all the files in the PR, thanks.

raphendyr commented on exercise/templates/exercise/staff/results.html:
```diff
@@ -10,45 +11,187 @@
+						title="{% trans 'Show unofficial points next to the official points in parentheses.' %}">
+						<input type="checkbox" class="unofficial-checkbox" value="unofficial">
+						{% trans "Show unofficial points" %}
+				  	</label>
```
This line contains space plus tabs. The spaces are not visible, as the tab masks them out.

raphendyr commented:
> Seems a bit unnecessary at this point, since teacher can still just unclick the checkbox?

The checkbox is pretty good. Though, it should only add brackets for those cells that actually have unofficial points. In addition, I would prefer if the text would be `Show only official points` and would be selected by default. Negation of official is not unofficial in this case as `official = not unofficial and not unconfirmed` or `official = ready and confirmed`.

raphendyr commented on exercise/staff_views.py:
```diff
@@ -179,7 +183,11 @@ class AllResultsView(CourseInstanceBaseView):
     def get_common_objects(self):
         super().get_common_objects()
         self.table = ResultTable(self.instance)
```
This can be dropped as it's not used. Am I right?

Mankro replied:
`ResultTable` builds the whole result table in the server side, i.e., fetches all submission data in the course from the database. Since the new results page fetches the data from the REST API, ResultTable is absolutely unnecessary in the view and must be removed.

raphendyr commented:
I think it's good enough for now. Ain't testing in production a thing?
